Introduce `profiles` (#98)
compose-spec/compose-spec: spec.md

As some Compose file elements can both be expressed as single strings or complex objects, merges MUST apply to
the expanded form.

### Profiles

Profiles allow to adjust the Compose application model for various usages and environments. A Compose
implementation SHOULD allow user to define a set of active profiles. The exact mechanism is implementation
specific and MAY include command line flags, environment variables, etc.

All top-level elements support a `profiles` attribute to define a list of named profiles. Elements without 
a `profiles` attribute set MUST always be enabled. The enclosing component MUST be ignored by compose
implementation when none of the listed profiles match the active ones, until the target component is

> explicitly targeted by a command.

Reviewer: It feels a bit strange that users can get (what is essentially) a one-off profile with always-on services + just the target service. This might get rather confusing since we're ignoring `profiles` attribute on the target service, but enforcing it everywhere else.

Is it worth to fomalise this a bit more by saying that `compose run/up TARGET` will activate the `profiles` of the target service? I appreciate that this probably makes the lives of Compose implementors harder (especially considering the follow-up profile-specific files proposal), but this functionality in isolation seems more confusing than beneficial.

> References to other services (by `links`, `extends` or shared resource syntax `service:xxx`) MUST not

Reviewer 2: See https://github.com/compose-spec/compose-spec/issues/97#issuecomment-685712571

Reviewer 3: The term "shared resource syntax `service:xxx`" is not defined anywhere in the spec document.

Where can a newcomer learn what is it? It would be good if there was at least a hyperlink here.

Reviewer 2: This is referring to system resources shareable between services or with the host, where you can specify the resource with the `service:[service name]` syntax, see [`network_mode`](https://github.com/compose-spec/compose-spec/blob/9015d1b3e0cd0504e38b9d22f8845068c7f22584/spec.md#network_mode) and [`ipc`](https://github.com/compose-spec/compose-spec/blob/9015d1b3e0cd0504e38b9d22f8845068c7f22584/spec.md#ipc) in the spec.

Reviewer 3: Thanks for the clarification. If the document included these links to the spec that would be really helpful for other people, too. Do you think it can be added?

Reviewer 3: Now I see that this snippet was already merged in https://github.com/compose-spec/compose-spec/pull/110/files#diff-bc6661da34ecae62fbe724bb93fd69b91a7f81143f2683a81163231de7e3b545R182.

My comment should really be placed there. The gist of it stands: could a link [network_mode](https://github.com/compose-spec/compose-spec/blob/9015d1b3e0cd0504e38b9d22f8845068c7f22584/spec.md#network_mode) and [ipc](https://github.com/compose-spec/compose-spec/blob/9015d1b3e0cd0504e38b9d22f8845068c7f22584/spec.md#ipc) be added in the document?

automatically enable a component that would otherwise have been ignored by active profiles.

> #### illustration sample

Reviewer: ```suggestion
#### Illustration sample
```
Since other headings uppercase first letter

```yaml

> services:
>   foo:
>      image: foo
>   bar:
>      image: bar
>      profiles: ["test"]
>   zot:
>      image: zot
>      profiles: ["debug"]
>      depends_on: ["bar"]

Reviewer 4: ```suggestion
services: 
  foo: 
    image: foo
  bar: 
    image: bar
    profiles: 
      - test
  zot: 
    depends_on: 
      - bar
    image: zot
    profiles: 
      - debug
```

```

- Compose application model parsed with no profile enabled only contains the `foo` service.
- If profile `test` is enabled, model contains both the `foo` and `bar` service.

> - If profile `debug` is enabled, model contains both the `foo` and `zot` service, but not `bar` 

Reviewer 4: ```suggestion
- If profile `debug` is enabled, model contains both `foo` and `zot` services, but not `bar` 
```

`privileged` configures the service container to run with elevated privileges. Support and actual impacts are platform-specific.

### profiles

> `profiles` defines a list of named profile for service to be enabled. When not set, service is always enabled.

Reviewer: ```suggestion
`profiles` defines a list of named profiles for service to be enabled under. When not set, service is always enabled.
```

Compose implementations MUST set `com.docker.compose.project` and `com.docker.compose.network` labels.

### profiles

> `profiles` defines a list of named profile for network to be enabled. When not set, network is always enabled

Reviewer: ```suggestion
`profiles` defines a list of named profiles for network to be enabled under. When not set, network is always enabled
```

      name: ${DATABASE_VOLUME}
```

### profiles

> `profiles` defines a list of named profile for volume to be enabled. When not set, volume is always enabled.

Reviewer: ```suggestion
`profiles` defines a list of named profiles for volume to be enabled under. When not set, volume is always enabled.
```
